### PR TITLE
more changes to compile darwin

### PR DESCRIPTION
**Pull request by baude**
this should represent the last major changes to get darwin to **compile**.  again,
the purpose here is to get darwin to compile so that we can eventually implement a
ci task that would protect against regressions for darwin compilation.

i have left the manual darwin compilation largely static still and in fact now only
interject (manually) two build tags to assist with the build.  trevor king has great
ideas on how to make this better and i will defer final implementation of those
to him.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: libpod/finished_64.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// +build !arm,!386
+// +build !arm,!386, linux
```
Comment (rhatdan) [inline libpod/finished_64.go]: Is the space before linux ok here?  Should this be 
// +build !arm,!386,linux

File: libpod/finished_unsupported.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// +build darwin, !linux
```
Comment (rhatdan) [inline libpod/finished_unsupported.go]: Ditto the space here.

**Comment by rhatdan** [comment]
Couple of nits on spaces, but LGTM
@wking @vbatts PTAL

**Comment by giuseppe** [comment]
LGTM

File: Makefile [hunk 3]
```diff
@@ -101,7 +101,7 @@ podman: .gopathok $(PODMAN_VARLINK_DEPENDENCIES)
 	$(GO) build -i -ldflags '$(LDFLAGS_PODMAN)' -tags "$(BUILDTAGS)" -o bin/$@ $(PROJECT)/cmd/podman
 
 darwin:
-	GOOS=darwin $(GO) build -ldflags '$(LDFLAGS_PODMAN)' -tags "$(BUILDTAGS)" -o bin/darwin$@ $(PROJECT)/cmd/podman
+	GOOS=darwin $(GO) build -ldflags '$(LDFLAGS_PODMAN)' -tags "containers_image_ostree_stub containers_image_openpgp" -o bin/podman.$@ $(PROJECT)/cmd/podman
```
Comment (mheon) [inline Makefile]: We should have a variable to hold the tags - BUILDTAGS_OSX or similar - so people can set them via environment variable when building

File: libpod/boltdb_state_linux.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	"encoding/json"
+	"path/filepath"
+
+	"github.com/boltdb/bolt"
+	"github.com/containers/storage"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+func (s *BoltState) setNamespace(ctr *Container, netNSPath string, newState *containerState) error {
```
Comment (mheon) [inline libpod/boltdb_state_linux.go]: This shouldn't be a BoltDB thing - the logic is generally just for handling network namespaces. Something like `replaceNetworkNamespace(ctr *Container, newNSPath string) error`

File: libpod/boltdb_state_linux.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+		// The container no longer has a network namespace
+		// Tear down the old one
+		if err := s.runtime.teardownNetNS(ctr); err != nil {
+			logrus.Warnf(err.Error())
+		}
+	}
+	return nil
+}
+
+func (s *BoltState) setNamespaceStatePath(ctr *Container) string {
```
Comment (mheon) [inline libpod/boltdb_state_linux.go]: This seems like a more general container thing - maybe add a `GetNetNSPath() string` to container.go?

File: libpod/boltdb_state_linux.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+	if ctr.state.NetNS != nil {
+		return ctr.state.NetNS.Path()
+	}
+	return ""
+}
+
+func (s *BoltState) getContainerFromDB(id []byte, ctr *Container, ctrsBkt *bolt.Bucket) error {
```
Comment (mheon) [inline libpod/boltdb_state_linux.go]: I really don't like ripping entire functions out. This doesn't seem at all cross-platform besides the netns code, and other than that I don't see any reason why we wouldn't share this cross-platform.

File: libpod/container_internal.go [hunk 7]
```diff
@@ -748,9 +744,8 @@ func (c *Container) mountStorage() (err error) {
 
 		if !mounted {
 			shmOptions := fmt.Sprintf("mode=1777,size=%d", c.config.ShmSize)
-			if err := unix.Mount("shm", c.config.ShmDir, "tmpfs", unix.MS_NOEXEC|unix.MS_NOSUID|unix.MS_NODEV,
-				label.FormatMountLabel(shmOptions, c.config.MountLabel)); err != nil {
-				return errors.Wrapf(err, "failed to mount shm tmpfs %q", c.config.ShmDir)
+			if err := c.platformMountStorage(shmOptions); err != nil {
```
Comment (mheon) [inline libpod/container_internal.go]: This isn't mounting storage - suggest renaming to `mountSHM` or similar.

File: libpod/container_internal.go [hunk 8]
```diff
@@ -786,53 +781,6 @@ func (c *Container) mountStorage() (err error) {
 	return c.save()
 }
 
-// prepare mounts the container and sets up other required resources like net
-// namespaces
-func (c *Container) prepare() (err error) {
```
Comment (mheon) [inline libpod/container_internal.go]: Again, why are we removing the entirety of these functions? I think all we really need here is to move the `c.config.CreateNetNS && c.state.NetNS == nil && c.config.PostConfigureNetNS` check into `createNetNS()` (and just have it return nil if that doesn't pass)

File: libpod/container_internal.go [hunk 9]
```diff
@@ -786,53 +781,6 @@ func (c *Container) mountStorage() (err error) {
...
-			return err
-		}
-	}
-
-	return nil
-}
-
-// cleanupNetwork unmounts and cleans up the container's network
-func (c *Container) cleanupNetwork() error {
-	if c.state.NetNS == nil {
```
Comment (mheon) [inline libpod/container_internal.go]: Same here - this is the only part that's really platform-specific, so we shouldn't move the entire function out

File: libpod/container_internal.go [hunk 10]
```diff
@@ -842,10 +790,8 @@ func (c *Container) cleanupStorage() error {
 	}
 
 	for _, mount := range c.config.Mounts {
-		if err := unix.Unmount(mount, unix.MNT_DETACH); err != nil {
-			if err != syscall.EINVAL {
-				logrus.Warnf("container %s failed to unmount %s : %v", c.ID(), mount, err)
-			}
+		if err := c.platformUnmountStorage(mount); err != nil {
```
Comment (mheon) [inline libpod/container_internal.go]: Suggest rename to unmountSHM

File: libpod/container_internal.go [hunk 11]
```diff
@@ -1178,173 +1124,6 @@ func (c *Container) generateHosts() (string, error) {
...
-		}
-		logrus.Debugf("Setting CGroup path for container %s to %s", c.ID(), cgroupPath)
-		g.SetLinuxCgroupsPath(cgroupPath)
-	}
-
-	return g.Config, nil
-}
-
 // Add an existing container's namespace to the spec
 func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr string, specNS string) error {
```
Comment (mheon) [inline libpod/container_internal.go]: If we move generateSpec into a platform-specific file, we should move this to the same place - it has no use other than the Linux version of generateSpec.

File: libpod/finished_64.go (hunk 1 above)
Comment (baude) [reply libpod/finished_64.go]: fixed

File: libpod/finished_unsupported.go (hunk 2 above)
Comment (baude) [reply libpod/finished_unsupported.go]: fixed

File: Makefile (hunk 3 above)
Comment (baude) [reply Makefile]: agree

Comment (wking) [reply Makefile]: > We should have a variable to hold the tags - BUILDTAGS_OSX...

Or pick conservative tags, like I've dond in #1034.

File: libpod/boltdb_state_linux.go (hunk 4 above)
Comment (baude) [reply libpod/boltdb_state_linux.go]: fixed per our discussion

File: libpod/boltdb_state_linux.go (hunk 5 above)
Comment (baude) [reply libpod/boltdb_state_linux.go]: fixed per our discussion

File: libpod/boltdb_state_linux.go (hunk 6 above)
Comment (baude) [reply libpod/boltdb_state_linux.go]: per our discussion, whoever does darwin|windows needs to parse this out.

File: libpod/container_internal.go (hunk 8 above)
Comment (baude) [reply libpod/container_internal.go]: per our discussion, whoever does darwin|windows needs to parse this out

File: libpod/container_internal.go (hunk 9 above)
Comment (baude) [reply libpod/container_internal.go]: per our discussion, whoever does darwin|windows needs to parse this out

File: Makefile (hunk 3 above)
Comment (baude) [reply Makefile]: @wking i assumming your PR will be rebased on mine and clean that up

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

File: libpod/finished_unsupported.go [hunk 12]
```diff
@@ -0,0 +1,12 @@
+// +build darwin,!linux
```
Comment (wking) [inline libpod/finished_unsupported.go]: Shouldn't this have been "everything not covered by `finished_32.go` and `finished_64.go`"?

File: libpod/finished_64.go [hunk 13]
```diff
@@ -1,4 +1,4 @@
-// +build !arm,!386
+// +build !arm,!386,linux
```
Comment (wking) [inline libpod/finished_64.go]: Shouldn't `finished_32.go` have gained a `,linux` too?

**Pull request closed without merging**